In [1]:
import pandas as pd
from sklearn.metrics import confusion_matrix
import numpy as np
import jsonlines
from termcolor import colored
from sklearn.metrics import classification_report

In [2]:
LABELS = ['B-MUS', 'B-PER', 'I-MUS', 'I-PER', 'O']    

original format:
```
{"id": 50, "sent_items": [["A", "O"], ["large", "O"], ["ensemble", "O"], ["of", "O"], ["trombonists", "O"], ["would", "O"], ["gather", "O"], ["to", "O"], ["play", "O"], ["music", "O"], ["written", "O"], ["for", "O"], ["multiple", "O"], ["trombones", "O"], ["or", "O"], ["transcribed", "O"], ["from", "O"], ["other", "O"], ["sources", "O"], [",", "O"], ["such", "O"], ["as", "O"], ["the", "O"], ["chorales", "O"], ["of", "O"], ["Johann", "B-MUS"], ["Sebastian", "I-MUS"], ["Bach", "I-MUS"], [".", "O"]]}
```

In [3]:
def convert_dev_from_json_to_conll(dev_path, conll_path):
    with jsonlines.open(dev_path, "r") as fin:
        with open(conll_path, "w") as fout:
            for line in fin:
                tokens = line["sent_items"]
                print(tokens)
                print([f"{t[0]} {t[1]}\n" for t in tokens])
                for t in tokens:
                    fout.write(f"{t[0]} {t[1]}\n")
                    
                    
def get_set_labels(set_path):
    set_labels = []
    with open(set_path, "r") as f:
        for t in f.readlines():
            if t.strip():
                label = t.split()[-1]
                set_labels.append(label)
            else:
                print(t)
    return set_labels
                    


In [4]:
dev_path = '../data/musicians_dataset/dev_converted.jsonl'
conll_path = '../experiments/gold_devs/dev_converted.conll'


convert_dev_from_json_to_conll(dev_path, conll_path)



In [8]:
gold_path = '../experiments/gold_devs/dev_converted.conll'

pred_path = '../experiments/predictions/sents_with_no_ents_bigger-manual.conll'
gold_labels = get_set_labels(gold_path)
pred_labels = get_set_labels(pred_path)
print(set(gold_labels))
print(set(pred_labels))


confusion_matrix(gold_labels, pred_labels, labels=['B-MUS', 'B-PER', 'I-MUS', 'I-PER', 'O'])

{'B-PER', 'I-MUS', 'O', 'I-PER', 'B-MUS'}
{'B-PER', 'I-MUS', 'O', 'I-PER', 'B-MUS'}


array([[  26,   62,    0,    1,    9],
       [   0,  162,    0,    6,   11],
       [   1,    0,   28,   57,    2],
       [   0,    0,    0,  106,    8],
       [   0,   26,    4,    9, 4028]])

In [14]:
dev_path = '../data/musicians_dataset/split_dev_20211118.jsonl'
conll_path = '../experiments/gold_devs/split_dev_20211118.conll'


convert_dev_from_json_to_conll(dev_path, conll_path)

[['In', 'O'], ['the', 'O'], ['pilot', 'O'], ['episode', 'O'], ['of', 'O'], ['the', 'O'], ['sequel', 'O'], ['series', 'O'], ['"', 'O'], ['Joey', 'O'], ['"', 'O'], [',', 'O'], ['Joey', 'B-PER'], ["'s", 'O'], ['sister', 'O'], ['Gina', 'B-PER'], ['briefly', 'O'], ['expresses', 'O'], ['her', 'O'], ['belief', 'O'], ['that', 'O'], ['Joey', 'B-PER'], ['and', 'O'], ['Chandler', 'B-PER'], ['were', 'O'], ['a', 'O'], ['gay', 'O'], ['couple', 'O'], ['.', 'O']]
['In O\n', 'the O\n', 'pilot O\n', 'episode O\n', 'of O\n', 'the O\n', 'sequel O\n', 'series O\n', '" O\n', 'Joey O\n', '" O\n', ', O\n', 'Joey B-PER\n', "'s O\n", 'sister O\n', 'Gina B-PER\n', 'briefly O\n', 'expresses O\n', 'her O\n', 'belief O\n', 'that O\n', 'Joey B-PER\n', 'and O\n', 'Chandler B-PER\n', 'were O\n', 'a O\n', 'gay O\n', 'couple O\n', '. O\n']
[['It', 'O'], ['was', 'O'], ['subsequently', 'O'], ['covered', 'O'], ['by', 'O'], ['many', 'O'], ['artists', 'O'], [',', 'O'], ['most', 'O'], ['famously', 'O'], ['by', 'O'], ['Quincy'

In [16]:
len(['The O\n', 'Polish O\n', 'NREAP O\n', 'plan O\n', 'is O\n', 'targeting O\n', '6,700 O\n', 'MW O\n', 'of O\n', 'wind O\n', 'power O\n', 'by O\n', '2020 O\n', 'whilst O\n', 'EWEA O\n', "'s O\n", '2009 O\n', 'forecast O\n', 'suggests O\n', 'a O\n', 'higher O\n', 'wind O\n', 'capacity O\n', 'of O\n', '10–12 O\n', '\xa0 O\n', 'GW O\n', 'is O\n', 'possible O\n', '. O\n'])

30

In [10]:
gold_path = '../experiments/gold_devs/split_dev_20211118.conll'

pred_path = '../experiments/predictions/sents_with_no_ents_bigger-split.conll'
gold_labels = get_set_labels(gold_path)
pred_labels = get_set_labels(pred_path)
print(set(gold_labels) - set(pred_labels))
print()


confusion_matrix(gold_labels, pred_labels, labels=LABELS)

set()



ValueError: Found input variables with inconsistent numbers of samples: [7677, 7662]

In [20]:
import re

def untag_dev_sentences(devpath):
    clean_sentences = []
    tokens = []
    with open(devpath, "r") as fin:
        for line in fin.readlines():
            sentence = re.sub('\-\[[A-Z]*\]',  '', line)
            clean_sentences.append(sentence)
            for token in sentence.split():
                if token.strip():
                    tokens.append(token)
    return clean_sentences, tokens

In [21]:
clean_sentences, tokens = untag_dev_sentences("../data/musicians_dataset/split_dev_2.jsonl")

In [23]:
tokens
# len(gold_labels)

['{"id":',
 '72163589,',
 '"sent_items":',
 '[["In",',
 '"O"],',
 '["the",',
 '"O"],',
 '["three",',
 '"O"],',
 '["following",',
 '"O"],',
 '["years",',
 '"O"],',
 '["she",',
 '"O"],',
 '["participated",',
 '"O"],',
 '["again",',
 '"O"],',
 '["in",',
 '"O"],',
 '["the",',
 '"O"],',
 '["Festival",',
 '"O"],',
 '["of",',
 '"O"],',
 '["Sanremo",',
 '"O"],',
 '[",",',
 '"O"],',
 '["winning",',
 '"O"],',
 '["the",',
 '"O"],',
 '["critic",',
 '"O"],',
 '["\'s",',
 '"O"],',
 '["award",',
 '"O"],',
 '["in",',
 '"O"],',
 '["1987",',
 '"O"],',
 '["with",',
 '"O"],',
 '["the",',
 '"O"],',
 '["song",',
 '"O"],',
 '["\\"",',
 '"O"],',
 '["Primo",',
 '"B-MUS"],',
 '["tango",',
 '"O"],',
 '["\\"",',
 '"O"],',
 '["(",',
 '"O"],',
 '["First",',
 '"O"],',
 '["Tango",',
 '"O"],',
 '[")",',
 '"O"],',
 '[".",',
 '"O"]]}',
 '{"id":',
 '3119,',
 '"sent_items":',
 '[["The",',
 '"O"],',
 '["watershed",',
 '"O"],',
 '["is",',
 '"O"],',
 '["bordered",',
 '"O"],',
 '["on",',
 '"O"],',
 '["the",',
 '"O"],',
 '["so

### manual

         ['B-MUS', 'B-PER', 'I-MUS', 'I-PER', 'O'] 
['B-MUS', [  29,   61,    0,    1,    7],
 'B-PER',[   1,   158,    0,    6,   14],
 'I-MUS'[    1,    0,    32,   51,    4],
 'I-PER',[   0,    0,    0,   106,    8],
 'O']    [   1,   29,    1,   15,   4021]
 

### Slit
 
 

In [19]:
from collections import Counter

In [43]:
dev_musicians = []
with jsonlines.open('../data/musicians_dataset/split_dev_only_hearst.jsonl') as f:
    for sentence in f:
        musician = ""
        for item in sentence["sent_items"]:
            if "-MUS" in item[1]:
                musician += item[0] + " "
            elif item[1] == "O" and musician != "":
                dev_musicians.append(musician[:-1])
                musician = ""

In [45]:
train_musicians = []
with jsonlines.open('../data/musicians_dataset/split_train_only_hearst_uniques.jsonl') as f:
    for sentence in f:
        musician = ""
        for item in sentence["sent_items"]:
            if "-MUS" in item[1]:
                musician += item[0] + " "
            elif item[1] == "O" and musician != "":
                train_musicians.append(musician[:-1])
                musician = ""

In [27]:
mandev_musicians = []
with jsonlines.open('../data/musicians_dataset/dev_converted.jsonl') as f:
    for sentence in f:
        musician = ""
        for item in sentence["sent_items"]:
            if "-MUS" in item[1]:
                musician += item[0] + " "
            elif item[1] == "O" and musician != "":
                mandev_musicians.append(musician[:-1])
                musician = ""

In [30]:
print(len(mandev_musicians), len(dev_musicians))

97 148


In [33]:
dev_musicians

['Don Fleming',
 'Airto Moreira',
 'Joe Farrell',
 'Duke Ellington',
 'Elliott Sharp',
 'Martin Luther',
 'Joseph Haydn',
 'Alfred Cellier',
 'Hélio Delmiro',
 'Maeve Mackinnon',
 'Miles Davis',
 'Franciscus Henricus Antheunis',
 'Elvis Costello',
 'Elvis Costello',
 'Gary Brooker',
 'Jerry Garcia',
 'Nicky Hopkins',
 'Steve Kimock',
 'Bob Weir',
 'Mickey Hart',
 'Richard Wagner',
 'Luciano Pavarotti',
 'Bob Dylan',
 'Dean Martin',
 'Frank Sinatra',
 'Tom Jones',
 'Michael Devlin',
 'Chick Corea',
 'Michael Jackson',
 'Thomas Martin Dumont',
 'Elton John',
 'Peter Gabriel',
 'Igor Stravinsky',
 'Bonnie Tyler',
 'Elvis Presley',
 'Pharrell Williams',
 'Rashied Ali',
 'Marilyn Crispell',
 'Matthijs Vermeulen',
 'Martin Luther',
 'Pierre Boulez',
 'Billy Joel',
 'Johanna Nurmimaa',
 'Warren Ellis',
 'Martin Luther',
 'Wynton Marsalis',
 'Luke Brandon',
 'Charles Aznavour',
 'Frank Sinatra',
 'Amélie Veille',
 'Jeff Anderson',
 'Martha Wash',
 'B.B. King',
 'Bernie Tormé',
 'Johann Sebasti

In [46]:
Counter(train_musicians).most_common(100)

[('Barbara Casini', 1),
 ('Marcus Byrne', 1),
 ('Landon Ronald', 1),
 ('Ramsey Lewis', 1),
 ('Jack Wilson', 1),
 ('Megan Washington', 1),
 ('Criss Oliva', 1),
 ('John Fayette Frame', 1),
 ('Bill Janovitz', 1),
 ('Maria Taylor', 1),
 ('Omar Jeffery Pineiro', 1),
 ('Joey Ramone', 1),
 ('Richard Bennett', 1),
 ('Wynton Marsalis', 1),
 ('Gus Cannon', 1),
 ('Paul Della Pelle', 1),
 ('Pepe Aguilar', 1),
 ('The Nicholas brothers', 1),
 ('Donald Nally', 1),
 ('Andreas Skår Winther', 1),
 ('Antonique Smith', 1),
 ('David Lowe', 1),
 ('Robert Kyagulanyi Ssentamu', 1),
 ('Fateh Ali Khan', 1),
 ('Charles Segal', 1),
 ('Al Foster', 1),
 ('Willem Mengelberg', 1),
 ('Alexander von Zemlinsky', 1),
 ('Oliver Korte', 1),
 ('Al Kooper', 1),
 ('Sigur Rós', 1),
 ("Colby O'Donis Colón", 1),
 ('Anne - Marie David', 1),
 ('Gennady Provatorov', 1),
 ('Guillermo Cides', 1),
 ('Jarvis Cocker of Pulp', 1),
 ('Issei Noro', 1),
 ('Albert Dadon AM', 1),
 ('Geddy Lee', 1),
 ('Rob Brown', 1),
 ('Joseph Hoffman', 1),
 

In [37]:
repetitions_in_dev = 0
repetitions_in_train = 0

mandev_musicians_set = set(mandev_musicians)
for x in Counter(train_musicians).most_common(100):
    if x[0] in mandev_musicians_set:
        print(x)
        repetitions_in_train += x[1]
        repetitions_in_dev += mandev_musicians.count(x[0])

print(repetitions_in_dev, repetitions_in_train)

('Paul McCartney', 63)
('Duke Ellington', 47)
('Johann Sebastian Bach', 40)
('Jools Holland', 22)
('Ella Fitzgerald', 22)
('Dave Matthews', 17)
6 211


In [39]:
print(len(dev_musicians), len(mandev_musicians))

148 97


In [44]:
repetitions_in_dev = 0
repetitions_in_train = 0

dev_musicians_set = set(dev_musicians)
for x in Counter(train_musicians).most_common(100):
    if x[0] in dev_musicians_set:
        print(x)
        repetitions_in_train += x[1]
        repetitions_in_dev += dev_musicians.count(x[0])

print(repetitions_in_dev, repetitions_in_train)

('Martin Luther', 143)
('Michael Jackson', 101)
('Elvis Presley', 92)
('Bob Dylan', 88)
('Paul McCartney', 63)
('Frank Sinatra', 58)
('Elton John', 49)
('Jennifer Lopez', 48)
('Duke Ellington', 47)
('Richard Wagner', 42)
('Johann Sebastian Bach', 40)
('Eric Clapton', 39)
('Miles Davis', 38)
('Britney Spears', 37)
('Stevie Wonder', 34)
('Celine Dion', 33)
('Dean Martin', 31)
('James Arthur', 30)
('John Coltrane', 27)
('Billy Joel', 24)
('Whitney Houston', 24)
('Elvis Costello', 23)
('Ray Charles', 22)
('John Cage', 22)
('Robert Schumann', 22)
('Buddy Holly', 21)
('Joseph Haydn', 21)
('Kelly Clarkson', 20)
('Henry Santos', 19)
('Chick Corea', 19)
('Tom Jones', 19)
('Muddy Waters', 19)
('James Taylor', 18)
('B.B. King', 18)
('Leonard Cohen', 17)
('Peter Gabriel', 16)
50 1384


In [42]:
len(dev_musicians)

79

In [26]:
dev_musicians_set = set(dev_musicians)
for x in Counter(train_musicians).most_common(100):
    if x[0] in dev_musicians_set:
        print(x)

('Martin Luther', 143)
('Michael Jackson', 101)
('Elvis Presley', 92)
('Bob Dylan', 88)
('Paul McCartney', 63)
('Frank Sinatra', 58)
('Elton John', 49)
('Jennifer Lopez', 48)
('Duke Ellington', 47)
('Richard Wagner', 42)
('Johann Sebastian Bach', 40)
('Eric Clapton', 39)
('Miles Davis', 38)
('Britney Spears', 37)
('Stevie Wonder', 34)
('Celine Dion', 33)
('Dean Martin', 31)
('James Arthur', 30)
('John Coltrane', 27)
('Billy Joel', 24)
('Whitney Houston', 24)
('Elvis Costello', 23)
('Ray Charles', 22)
('John Cage', 22)
('Robert Schumann', 22)
('Buddy Holly', 21)
('Joseph Haydn', 21)
('Kelly Clarkson', 20)
('Henry Santos', 19)
('Chick Corea', 19)
('Tom Jones', 19)
('Muddy Waters', 19)
('James Taylor', 18)
('B.B. King', 18)
('Leonard Cohen', 17)
('Peter Gabriel', 16)


In [2]:
PREDICTIONS_PATH = "../experiments/predictions"

In [11]:
with open(f"{PREDICTIONS_PATH}/only_hearst_uniques-manual.conll", 'r') as f:
    sentence = []
    for line in f.readlines():
        word, tag = line.strip().split(" ", 1)
        if word == ".":
            color_text = [x[0] if x[1] == "O" else colored(x[0], 'red') if "PER" in x[1] else colored(x[0], 'green') for x in sentence]
            print(" ".join(color_text))
            print()
            sentence = []
        else:
            sentence.append([word, tag])


The General - Director of Rustavi 2 , Nika Tabatadze , denied that Khoperia , was under pressure from the Rustavi 2 management or the authorities

He said that instead of Bibiluri , Oleg Melnikov , was the fourth attacker

A large ensemble of trombonists would gather to play music written for multiple trombones or transcribed from other sources , such as the chorales of Johann Sebastian Bach

The sketch included appearances by then UK Prime Minister Gordon Brown , JLS , Paul McCartney and Justin Bieber

The band performed two songs on " Later ... with Jools Holland " during May 2009 and have supported Blur at Hyde Park , London

Economist John Quiggin distinguishes between the Laffer curve and Laffer 's analysis of tax rates

It was executed in 1513 as shown on the work itself - Francesco Tassi believed it was inscribed " NOB.PAULUS , ET JO

FRATRES DE CASPOTTIS TRINO OBTULERUNT HAEC 1513 "

The United States Supreme Court eventually affirmed the judgment of the New York Court of Appea

In [10]:
with jsonlines.open(f"../data/musicians_dataset/split_dev_only_hearst_uniques.jsonl", 'r') as f:
    for i, line in enumerate(f):
        color_text = [x[0] if x[1] == "O" else colored(x[0], 'red') if "PER" in x[1] else colored(x[0], 'green') for x in line["sent_items"]]
        print(i, line["id"], " ".join(color_text))
        print()

0 782602 Roger Tiegs ( born 18 June 1972 in Nordfjord ) , better known by his stage name Infernus , is a Norwegian black metal musician and Satanist .

1 242052 Each year , after setting choreography for the Ice Capades ’ season , Fletcher regularly choreographed shows in Las Vegas at the Desert Inn , Moulin Rouge , Latin Quarter , The Dunes and The Flamingo hotels for the likes of Pearl Bailey , Vic Damone , Shirley Jones , Paula Kelly , Tallulah Bankhead and Jane Kean ( with whom he also partnered on stage ) .

2 8234 In 1993 the bridge was rehabilitated by Milton Graton Associates , after the Vermont Agency of Transportation ( AOT ) found it to be inadequate for normal loading .

3 59593 Manda discovers that Darren Miller ( Charlie G. Hawkins ) is secretly the father of Heather Trott ( Cheryl Fergison ) 's baby George , but agrees to keep it to herself .

4 66069 Between 1970 and 1975 , Lockheed paid Saudi Arms dealer Adnan Khashoggi $ 106   million in commissions .

5 31425 Feliz m

In [5]:
bad_df = pd.read_csv("../tests/fixtures/bad_model_gold_and_predictions.tsv", sep="\t", names=["gold", "pred"])

df["match"] = "_"

In [7]:
df.to_csv("../experiments/predictions/hit_or_miss.csv", sep=" ")

In [6]:
golds = df["gold"].tolist()
preds = df["pred"].tolist()

classification_report(golds, preds, labels=["B-PER", "B-MUS", "I-PER", "I-MUS", "O"])
            
        

'              precision    recall  f1-score   support\n\n       B-PER       0.80      0.86      0.83       613\n       B-MUS       0.77      0.30      0.43       115\n       I-PER       0.78      0.91      0.84       537\n       I-MUS       0.81      0.32      0.46       134\n           O       0.99      0.99      0.99      7982\n\n    accuracy                           0.96      9381\n   macro avg       0.83      0.68      0.71      9381\nweighted avg       0.96      0.96      0.95      9381\n'

In [16]:
def tag_hits(df, pos, neg):
    neg_labels = ["O", f"B-{neg}", f"I-{neg}"]
    for i, row in df.iterrows():
        if row["match"] == "_":
            if (row["gold"] in neg_labels) and (row["pred"] in neg_labels):
                df.at[i,'match'] = "tn"
            elif (row["gold"] in neg_labels) and (row["pred"].endswith(pos)):
                df.at[i,'match'] = "fp"
            elif (row["gold"].endswith(pos)) and not (row["pred"].endswith(pos)):
                df.at[i,'match'] = "fn"
            elif (row["gold"] == f"B-{pos}") and (row["pred"] == f"B-{pos}"):
                span = []
                x = i+1
                error = False
                while df.at[x, "gold"] == f"I-{pos}":
                    span.append(x)
                    x += 1
                    if df.at[x, "gold"] != df.at[x, "pred"]:
                        error = True                    
                if error:
                    df.at[i,'match'] = "fn"
                else:
                    df.at[i,'match'] = "tp"
                for x in span:
                    df.at[x,'match'] = "--"


def get_span_recall(filepath, positive_label, negative_label):
    labels = [f"B-{positive_label}", f"I-{positive_label}", f"B-{negative_label}", f"I-{negative_label}", "O"]
    df = pd.read_csv(filepath, sep="\t", names=["gold", "pred"])
    golds = df["gold"].tolist()
    preds = df["pred"].tolist()
    report  = classification_report(golds, preds)
    print(report)
    df["match"] = "_"
    tag_hits(df, positive_label, negative_label)
    print("true positive spans:", df[df["match"] == "tp"].count())
    print("total gold positives:", df[df["gold"] == f"B-{positive_label}"].count())

In [17]:
get_span_recall("../tests/fixtures/bad_model_gold_and_predictions.tsv", "MUS", "PER")

              precision    recall  f1-score   support

       B-MUS       1.00      0.75      0.86         4
       B-PER       0.90      0.75      0.82        12
       I-MUS       0.75      0.60      0.67         5
       I-PER       0.90      1.00      0.95         9
           O       0.98      0.99      0.99       170

    accuracy                           0.96       200
   macro avg       0.91      0.82      0.85       200
weighted avg       0.96      0.96      0.96       200

true positive spans: gold     1
pred     1
match    1
dtype: int64
total gold positives: gold     4
pred     4
match    4
dtype: int64


In [18]:
get_span_recall("../tests/fixtures/good_model_gold_and_predictions.tsv", "MUS", "PER")

              precision    recall  f1-score   support

       B-MUS       1.00      0.75      0.86         4
       B-PER       0.90      0.75      0.82        12
       I-MUS       1.00      0.75      0.86         4
       I-PER       0.90      1.00      0.95         9
           O       0.98      1.00      0.99       171

    accuracy                           0.97       200
   macro avg       0.96      0.85      0.89       200
weighted avg       0.97      0.97      0.97       200

true positive spans: gold     3
pred     3
match    3
dtype: int64
total gold positives: gold     4
pred     4
match    4
dtype: int64


In [9]:
df[df["match"] == "tp"].count()

gold     34
pred     34
match    34
dtype: int64

In [16]:
df[df["gold"] == "B-MUS"].count()

gold     115
pred     115
match    115
dtype: int64

In [17]:
34/115*100

29.565217391304348

In [6]:
for i, row in df.iterrows():
    if row["match"] == "_":
        if (row["gold"] in ["O", "B-PER", "I-PER"]) and (row["pred"] in ["O", "B-PER", "I-PER"]):
            df.at[i,'match'] = "tn"
        elif (row["gold"] in ["O", "B-PER", "I-PER"]) and (row["pred"].endswith("MUS")):
            df.at[i,'match'] = "fp"
        elif (row["gold"].endswith("MUS")) and not (row["pred"].endswith("MUS")):
            df.at[i,'match'] = "fn"
        elif (row["gold"] == "B-MUS") and (row["pred"] == "B-MUS"):
            span = []
            x = i+1
            error = False
            while df.at[x, "gold"] == "I-MUS":
                span.append(x)
                x += 1
                if df.at[x, "gold"] != df.at[x, "pred"]:
                    print(df.at[x, :])
                    error = True                    
            if error:
                df.at[i,'match'] = "fn"
            else:
                df.at[i,'match'] = "tp"
            for x in span:
                df.at[x,'match'] = "--"
        
        

In [ ]:
results = []
for x in range(len(golds)):
    if golds[x] == "O" and preds[x] == "O":
        results.append(("O", "tp"))
    elif golds[x] == "O" and preds[x] != "O":
        results.append(("O", "fp"))
        

'        precision    recall  f1-score   support\n\n
B-PER       0.80      0.86      0.83       613\n  
B-MUS       0.77      0.30      0.43       115\n       
I-PER       0.78      0.91      0.84       537\n       
I-MUS       0.81      0.32      0.46       134\n
O           0.99      0.99      0.99      7982\n\n    

accuracy                        0.96      9381\n   
macro avg   0.83      0.68      0.71      9381\n
weighted avg 0.96      0.96      0.95      9381\n'